### PR TITLE
refactor(predployedContracts): add constants support

### LINKED ISSUE #237: Improve ABIs and predeployed contracts
## Issue Overview

create a `constants.ts` file that stores the addresses & classhash of `Eth`, `Strk`, and one single `ERC20 ABI`. we should be able to use these inside `predeployedContracts.ts` 

all these values are constant and are the same regardless of the network.

also add support of `Strk`, `Eth` on mainnet.

## Proposed Solutions or Ideas

currently ABI in devnet doesnt match ABI on sepolia, due to a previous mismatch between devnet and public networks. But now this should be solved. use ONE single ABI for all tokens.

### PR DESCRIPTION
**Pull request by oceanlvr**
# Task name here

Fixes #237 

## Types of change

- [ ] Feature
- [ ] Bug
- [x] Enhancement

## Comments (optional)

> Issue Overview
create a constants.ts file that stores the addresses & classhash of Eth, Strk, and one single ERC20 ABI. we should be able to use these inside predeployedContracts.ts
all these values are constant and are the same regardless of the network.
also add support of Strk, Eth on mainnet.
Proposed Solutions or Ideas
currently ABI in devnet doesnt match ABI on sepolia, due to a previous mismatch between devnet and public networks. But now this should be solved. use ONE single ABI for all tokens.

Added constants that can be used for predeployedContract.

### CODE REVIEW

File: packages/nextjs/contracts/Constants.ts [hunk 1]
```diff
@@ -0,0 +1,218 @@
+const universalEthAddress =
+  "0x49D36570D4E46F48E99674BD3FCC84644DDD6B96F7C741B1562B82F9E004DC7" as const;
+
+const universalStrkAddress =
+  "0x04718f5a0fc34cc1af16a1cdee98ffb20c31f5cd61d6ab07201858f4287c938d" as const;
+
+const universalClassHash = '0x046ded64ae2dead6448e247234bab192a9c483644395b66f2155f2614e5804b0' as const;
+
+const universalAbi = [
```
Comment (oceanlvr) [inline packages/nextjs/contracts/Constants.ts:9]: > currently ABI in devnet doesnt match ABI on sepolia, due to a previous mismatch between devnet and public networks. But now this should be solved. use ONE single ABI for all tokens.

sir, we should have a unified ABI here, but the ABI configuration in devnet is different from that in Sepolia and other networks. Is the code here using Sepolia's ABI or devnet's ABI?

File: packages/nextjs/contracts/predeployedContracts.ts [hunk 2]
```diff
@@ -2,951 +2,47 @@
...
+      address: universalEthAddress,
+      abi: universalAbi,
+      classHash: universalClassHash,
+    },
+    Strk: {
+      address: universalStrkAddress,
+      abi: universalAbi,
+      classHash: universalClassHash,
+    },
+  }
```
Comment (oceanlvr) [inline packages/nextjs/contracts/predeployedContracts.ts:45]: add eth & strk for mainnet, Are these values ​​correct? (keep it same as devnet)

**Comment by 0xquantum3labs** [comment]
@oceanlvr you can try checking whether the ABIs of ERC20 in sepolia and devnet are same. [here](https://docs.openzeppelin.com/contracts-cairo/0.14.0/erc20#interface) is the ERC20 standard. 

this are the main functions that have to be available on the ABI. If i remember correctly, previously devnet was missing one or two of the CamelCase functions. This should've been resolved in latest versions of devnet, so now we should be able to have same ABI for both devnet and sepolia.

**Comment by oceanlvr** [comment]
sir i update this pr

> you can try checking whether the ABIs of ERC20 in sepolia and devnet are same

i use the `import { useProvider } from "@starknet-react/core";` API and find that sepolia-eth is not same as devnet-eth.

**Comment by 0xquantum3labs** [comment]
> sir i update this pr
> 
> > you can try checking whether the ABIs of ERC20 in sepolia and devnet are same
> 
> i use the `import { useProvider } from "@starknet-react/core";` API and find that sepolia-eth is not same as devnet-eth.

yes, devnet is different to sepolia and mainnet.

**Comment by oceanlvr** [comment]
PTAL sir @0xquantum3labs

File: packages/nextjs/contracts/Constants.ts
[the remark this answers is not in the document]
Comment (0xquantum3labs) [reply packages/nextjs/contracts/Constants.ts:1]: @oceanlvr

**Review by 0xquantum3labs — CHANGES_REQUESTED** [review]

**Comment by 0xquantum3labs** [comment]
@Nadai2010 can you provide the addresses & class hashes of Strk and Eth on devnet , sepolia, mainnet ?

**Comment by Nadai2010** [comment]
universalEthAddress: "0x49D36570D4E46F48E99674BD3FCC84644DDD6B96F7C741B1562B82F9E004DC7"

classHash Devnet:    "0x046ded64ae2dead6448e247234bab192a9c483644395b66f2155f2614e5804b0"

classHash Sepolia: "0x07f3777c99f3700505ea966676aac4a0d692c2a9f5e667f4c606b51ca1dd3420"

classHash Mainnet:
"0x07f3777c99f3700505ea966676aac4a0d692c2a9f5e667f4c606b51ca1dd3420"

universalStrkAddress = "0x04718f5a0fc34cc1af16a1cdee98ffb20c31f5cd61d6ab07201858f4287c938d"

classHash Devnet:        "0x046ded64ae2dead6448e247234bab192a9c483644395b66f2155f2614e5804b0"

classHash Sepolia: "0x04ad3c1dc8413453db314497945b6903e1c766495a1e60492d44da9c2a986e4b"

classHash Mainnet:
"0x05327164fa21dca89a92e8eae8a5b7ab90f58373e71f0a16d285e5a4abe5a3cf"

**Comment by oceanlvr** [comment]
sir, move file to `utils` fold & update all constants value.

**Comment by Nadai2010** [comment]
Sorry the classHash in mainnetSTRK is [0x04ad3c1dc8413453db314497945b6903e1c766495a1e60492d44da9c2a986e4b](https://starkscan.co/contract/0x04718f5a0fc34cc1af16a1cdee98ffb20c31f5cd61d6ab07201858f4287c938d)

The classHash for both tokens in Sepolia and Mainnent are the same, only in devnet they are different

```bash
// Eth
const sepoliaEthClassHash =
  "0x07f3777c99f3700505ea966676aac4a0d692c2a9f5e667f4c606b51ca1dd3420" as const;

const mainnetEthClassHash =
  "0x07f3777c99f3700505ea966676aac4a0d692c2a9f5e667f4c606b51ca1dd3420" as const;

// Strk
const sepoliaStrkClassHash =
  "0x04ad3c1dc8413453db314497945b6903e1c766495a1e60492d44da9c2a986e4b" as const;

const mainnetStrkClassHash =
  "0x04ad3c1dc8413453db314497945b6903e1c766495a1e60492d44da9c2a986e4b" as const;
```

----

We can leave `sepoliamainnetEthClassHash` and `sepoliamainnetStrkClassHash` example:

```bash
// Eth
const sepoliamainnetEthClassHash =
  "0x07f3777c99f3700505ea966676aac4a0d692c2a9f5e667f4c606b51ca1dd3420" as const;

// Strk
const sepoliamainnetStrkClassHash =
  "0x04ad3c1dc8413453db314497945b6903e1c766495a1e60492d44da9c2a986e4b" as const;
```

What do you think? @0xquantum3labs @jrcarlos2000

**Review by Nadai2010 — APPROVED** [review]

File: packages/nextjs/utils/Constants.ts [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+const universalStrkAddress =
+  "0x04718f5a0fc34cc1af16a1cdee98ffb20c31f5cd61d6ab07201858f4287c938d" as const;
+
+const sepoliaMainnetStrkClassHash =
+  "0x04ad3c1dc8413453db314497945b6903e1c766495a1e60492d44da9c2a986e4b" as const;
+
+const devnetStrkClassHash =
+  "0x046ded64ae2dead6448e247234bab192a9c483644395b66f2155f2614e5804b0" as const;
+
+const devnetAbi = [
```
Comment (jrcarlos2000) [inline packages/nextjs/utils/Constants.ts:21]: lets remove the `devnetAbi`, and use the `sepoliaAbi` as universal one and call it `universalErc20Abi` @oceanlvr @Nadai2010

**Review by jrcarlos2000 — CHANGES_REQUESTED** [review]

File: packages/nextjs/utils/Constants.ts (hunk 3 above)
Comment (oceanlvr) [reply packages/nextjs/utils/Constants.ts]: sir removed devnetAbi and use universalErc20Abi

**Review by Nadai2010 — APPROVED** [review]

**Pull request merged**